src/Sulu/Bundle/WebsiteBundle/Routing/ContentRouteProvider.php

```diff
@@ -61,224 +59,157 @@ class ContentRouteProvider implements RouteProviderInterface
+        // just show the page
+        $portal = $attributes->getAttribute('portal');
+        $locale = $attributes->getAttribute('localization')->getLocale();
+        $resourceLocatorStrategy = $this->resourceLocatorStrategyPool->getStrategyByWebspaceKey(
+            $portal->getWebspace()->getKey()
+        );
 
-                    if ($request->getQueryString()) {
-                        $redirectUrl .= '?' . $request->getQueryString();
-                    }
+        try {
+            // load content by url ignore ending trailing slash
+            /** @var PageDocument $document */
+            $document = $this->documentManager->find(
+                $resourceLocatorStrategy->loadByResourceLocator(
+                    rtrim($resourceLocator, '/'),
+                    $portal->getWebspace()->getKey(),
+                    $locale
+                ),
+                $locale,
+                [
+                    'load_ghost_content' => false,
+                ]
+            );
 
-                    $collection->add(
-                        $document->getStructureType() . '_' . $document->getUuid(),
-                        $this->getRedirectRoute(
-                            $request,
-                            $redirectUrl
-                        )
-                    );
-                } elseif ($document->getRedirectType() === RedirectType::EXTERNAL) {
-                    $collection->add(
-                        $document->getStructureType() . '_' . $document->getUuid(),
-                        $this->getRedirectRoute($request, $document->getRedirectExternal())
-                    );
-                } elseif (!$this->checkResourceLocator()) {
-                    return $collection;
-                } else {
-                    // convert the page to a StructureBridge because of BC
-                    $metadata = $this->documentInspector->getStructureMetadata($document);
-                    if (!$metadata) {
-                        return $collection;
-                    }
+            if (!$document->getTitle()) {
+                // If the title is empty the document does not exist in this locale
+                // Necessary because of https://github.com/sulu/sulu/issues/2724, otherwise locale could be checked
+                return $collection;
+            }
 
-                    /** @var PageBridge $structure */
-                    $structure = $this->structureManager->wrapStructure(
-                        $this->documentInspector->getMetadata($document)->getAlias(),
-                        $metadata
-                    );
-                    $structure->setDocument($document);
+            if (preg_match('/\/$/', $resourceLocator) && $prefix) {
+                // redirect page to page without slash at the end
+                $url = $prefix . rtrim($resourceLocator, '/');
+                $collection->add('redirect_' . uniqid(), $this->getRedirectRoute($request, $url));
```

Comment:
this redirect will be done here because the page is already loaded and the provider knows that the target exists.